### ISSUE TITLE
Update Install Steps for Linux, WSL, and iOS on Wiki

### ISSUE THREAD

**Issue by UsrLightmann** [theos/theos#551]
(Thanks for reporting an issue! Please make sure you click the link above to view the issue guidelines, then fill out the blanks below.)

What are the steps to reproduce this issue?
-------------------------------------------
Try to compile a project on Linux/WSL after following the steps currently outlined on the Linux/WSL install page on the Wiki.

What happens?
-------------
A whole host of errors (see below)

What were you expecting to happen?
----------------------------------
The project to compile without error(s).

Any logs, error output, etc?
----------------------------
After setting up a fresh Ubuntu install, following the steps outlined in the Wiki to install Theos, and trying to build a project for arm64/arm64e I encountered the following errors:

From step 5: 
- “Command 'unzip' not found, …”

Missing Dependency: 
- “/home/lightmann/theos/toolchain/linux/iphone/bin/arm64-apple-darwin14-ld: error while loading shared libraries: libtinfo.so.5: cannot open shared object file: No such file or directory”

Missing arm64e toolchain support:
- “ld: malformed file
/home/lightmann/theos/vendor/lib/CydiaSubstrate.framework/CydiaSubstrate.tbd:2:42: error: unknown bit value
archs:           [ armv7, armv7s, arm64, arm64e, i386, x86_64 ]
                                         ^~~~~~
 file '/home/lightmann/theos/vendor/lib/CydiaSubstrate.framework/CydiaSubstrate.tbd'
clang: error: linker command failed with exit code 1 (use -v to see invocation)”

Trying to compile a project for arm64e: 
- “clang: error: invalid arch name '-arch arm64e'”

Some users, myself included, have also encountered various errors from toolchain files because of the prefixes present on Kabir’s toolchain (and other arm64e toolchains online that also contain, or tell you to add, prefixes). 

- Ex: ”/usr/bin/ld: unrecognised emulation mode: ultiply_defined 
Supported emulations: elf_x86_64 elf32_x86_64 elf_i386 elf_iamcu elf_l1om elf_k1om i386pep i386pe 
clang-10: error: linker command failed with exit code 1 (use -v to see invocation)” 

When trying to compile a tweak on the C: drive: 
- “ERROR: control directory has bad permissions 777 (must be >=0755 and <=0775)” 

**To Fix:**
•	Install libtinfo5 along with the other prerequisite packages
•	Use Bingner’s arm64e linux toolchain in place of Kabir’s arm64 linux toolchain (no prefixes and arm64e support)
•	Create a wsl.conf file in roots/etc/ -- see [issue opened by Muirey03; he outlines the underlying cause and a fix](https://github.com/theos/theos/issues/421)

**What to update on the Wiki:**
Remove “Follow the instructions at http://apt.llvm.org to add the correct clang-6.0 source for your Linux distro. Then run sudo apt-get update to refresh your sources.” as it is not necessary + remove clang-6.0 from prereq install command below.

Step 1 – add libtinfo5 to install list (not required for some distros, however, as it is bundled with ncurses):
 `sudo apt-get install fakeroot git perl build-essential libtinfo5`

Additionally to fix permission error when working in the windows partition on WSL (not necessary if a fileserver or CLI are used instead): 
“Setup fakeroot: `sudo update-alternatives --set fakeroot /usr/bin/fakeroot-tcp
`
Create a text file in roots/etc/, name it wsl.conf, add the following contents:
```
[automount]
options="metadata"
```
Then restart WSL (LxssManager service).”

Step 4 – switch to Bingner’s arm64e toolchain:
```
curl -LO https://github.com/sbingner/llvm-project/releases/download/v10.0.0-1/linux-ios-arm64e-clang-toolchain.tar.lzma
TMP=$(mktemp -d)
tar --lzma -xvf linux-ios-arm64e-clang-toolchain.tar.lzma -C $TMP
mkdir -p $THEOS/toolchain/linux/iphone
mv $TMP/ios-arm64e-clang-toolchain/* $THEOS/toolchain/linux/iphone/
rm -r $TMP linux-ios-arm64e-clang-toolchain.tar.lzma
```

Any other comments?
-------------------
_These steps were tested on WSL with a fresh install of Ubuntu ([from the Microsoft Store)](https://www.microsoft.com/store/productId/9NBLGGH4MSV6)_

In summary: 
o	Removed unneeded step and respective package
o	Added missing dependency to prerequisite install command
o	Added a WSL specific step to allow for projects to be packaged on the C: drive
o	Switched to Bingner’s toolchain (supports arm64e and lacks prefixes)

What versions of software are you using?
----------------------------------------
**Operating System:** … N/A

**Toolchain Version:** … N/A

**SDK Version:** … N/A

**Comment by kabiroberai** [comment]
I agree that Sam's toolchain should be used instead of mine now, the existence of his toolchain being the reason I stopped updating mine. I don't have a WSL or Linux environment readily available atm though, so I'll defer to other maintainers to confirm that all of this works fine and to update the wiki page.

**Comment by kabiroberai** [comment]
I'm merging #517 into this and updating the title to reflect that the iOS instructions should be updated as well. As per @CRKatri:

> The wiki pages for both Linux and iOS recommend outdated toolchains incapable of compiling for arm64e.
>
> For iOS, the ideal solution is to create a meta-package containing these depends:
clang, ld64, odcctools, make, ldid, git
This will take the place of theos dependencies package and download an arm64e compatible toolchain from elucubratus, however, this will not work with the chimera bootstrap and am unsure if that is possible to achieve.
>
> For Linux the code snippet for downloading the toolchain should be replaced by
>
> ```
> curl https://github.com/sbingner/llvm-project/releases/download/v10.0.0-1/linux-ios-arm64e-clang-toolchain.tar.lzma -Lo toolchain.tar.lzma
> mkdir -p $THEOS/toolchain/linux/iphone/
> tar xf toolchain.tar.lzma -C $THEOS/toolchain/linux/iphone
> rm toolchain.tar.lzma
> ```

**Comment by DGh0st** [comment]
Few things to add:
- `libtinfo5` package only exists for some distros, on other distros its part of ncurses.
- Since `clang-10` is included in Bingner's toolchain, `clang-6.0` should also be removed from Step 1 command
- The WSL steps suggested is correct if user is creating projects on a mounted drive. But those steps are useless if WSL's P9 fileserver is used instead or just using CLI within the linux filesystem (each of those have benefits and flaws).

**Comment by UsrLightmann** [comment]
>     * `libtinfo5` package only exists for some distros, on other distros its part of ncurses.

>     * Since `clang-10` is included in Bingner's toolchain, `clang-6.0` should also be removed from Step 1 command

>     * The WSL steps suggested is correct if user is creating projects on a mounted drive. But those steps are useless if WSL's P9 fileserver is used instead or just using CLI within the linux filesystem (each of those have benefits and flaws).

Thanks for catching all of these! Updated the post accordingly.

**Comment by medoosh** [comment]
🤙🏼

**Comment by EthanRDoesMC** [comment]
dude you're freakin coming in clutch for me _again_

**Comment by CharlieWhile13** [comment]
cum <@&811493950443225088>

**Comment by tale** [comment]
Just thought I would mention this but it seems like @CRKatri has a toolchain https://github.com/CRKatri/llvm-project/releases/tag/swift-5.3.2-RELEASE. This supports Swift 5.3 and Clang 10. Should consider that too.

**Comment by UsrLightmann** [comment]
Little update: Bingner's toolchain **doesn't** work with 14 sdks:
```
ld: unsupported tapi file type '!tapi-tbd' in YAML file '/home/lightmann/theos/sdks/iPhoneOS14.4.sdk/usr/lib/libobjc.tbd' for architecture arm64e
clang-10: error: linker command failed with exit code 1 (use -v to see invocation)
make[3]: *** [/home/lightmann/theos/makefiles/instance/library.mk:52: /home/lightmann/quorra/.theos/obj/arm64e/Quorra.dylib] Error 1
```

I decided to try out CRKatri's toolchain (which tale linked above) and it **does** work with 14 sdks. I also tried with a 13 sdk and it worked there as well. 


Tested on Ubuntu 20.04 with the Ubuntu 18.04 tarball.


Think this may be the toolchain to link to from now on since PoomSmart has updated the [official sdks repo](https://github.com/theos/sdks) (where the latest sdk is 14.4).

@kabiroberai

**Comment by UsrLightmann** [comment]
Worth noting that with CRKatri's toolchain, libtinfo5 is no longer required and zstd is. 

With the 20.04 tarball specifically, libz3-dev is required (for the swift toolchain). I don't think the 18.04 tarball requires it and haven't tested the buster one.

**Comment by CRKatri** [comment]
I am contemplating packaging my toolchain as a deb as well, but still gotta get Swift building with dpkg-buildpackage, which couldn't be hard, but ideally installing the toolchain will be as simple as adding my repo and installing the toolchain. I will hopefully get it up this week.

**Comment by sbingner** [comment]
> Little update: Bingner's toolchain **doesn't** work with 14 sdks:
> 
> ```
> ld: unsupported tapi file type '!tapi-tbd' in YAML file '/home/lightmann/theos/sdks/iPhoneOS14.4.sdk/usr/lib/libobjc.tbd' for architecture arm64e
> clang-10: error: linker command failed with exit code 1 (use -v to see invocation)
> make[3]: *** [/home/lightmann/theos/makefiles/instance/library.mk:52: /home/lightmann/quorra/.theos/obj/arm64e/Quorra.dylib] Error 1
> ```
> 
> I decided to try out CRKatri's toolchain (which tale linked above) and it **does** work with 14 sdks. I also tried with a 13 sdk and it worked there as well.
> 
> Tested on Ubuntu 20.04 with the Ubuntu 18.04 tarball.
> 
> Think this may be the toolchain to link to from now on since PoomSmart has updated the [official sdks repo](https://github.com/theos/sdks) (where the latest sdk is 14.4).
> 
> @kabiroberai

That has been fixed - let me know if you have issues with ios14 sdks still

**Comment by sbingner** [comment]
https://github.com/sbingner/llvm-project/releases/tag/v10.0.0-2 should be good for linux with iOS14 SDKs - also updated cctools and ldid

**Comment by UsrLightmann** [comment]
> 
> 
> https://github.com/sbingner/llvm-project/releases/tag/v10.0.0-2 should be good for linux with iOS14 SDKs - also updated cctools and ldid

Everything seems good except ldid, which is acting up for me.

```
bash: /home/lightmann/theos/toolchain/linux/iphone/bin/ldid: cannot execute binary file: Exec format error
make[2]: *** [/home/lightmann/theos/makefiles/instance/library.mk:51: /home/lightmann/Documents/Tweaks/DEVELOPED/quorra/.theos/obj/Quorra.dylib] Error 126
```

```
file $THEOS/toolchain/linux/iphone/bin/ldid
/home/lightmann/theos/toolchain/linux/iphone/bin/ldid: ELF 64-bit LSB shared object, x86-64, version 1 (GNU/Linux), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=2971186340cb12b1a451da2301b69bc85a7b8d2a, for GNU/Linux 3.2.0, not stripped
```

```
uname -p -o 
x86_64 GNU/Linux
```

Not sure what's up. Had a friend test it and it works fine for them (w a 12 sdk as well as 14 sdk). We're both on Ubuntu 20.04 (WSL2).

**Comment by Thog** [comment]
> 
> 
> > https://github.com/sbingner/llvm-project/releases/tag/v10.0.0-2 should be good for linux with iOS14 SDKs - also updated cctools and ldid
> 
> Everything seems good except ldid, which is acting up for me.
> 
> ```
> bash: /home/lightmann/theos/toolchain/linux/iphone/bin/ldid: cannot execute binary file: Exec format error
> make[2]: *** [/home/lightmann/theos/makefiles/instance/library.mk:51: /home/lightmann/Documents/Tweaks/DEVELOPED/quorra/.theos/obj/Quorra.dylib] Error 126
> ```
> 
> ```
> file $THEOS/toolchain/linux/iphone/bin/ldid
> /home/lightmann/theos/toolchain/linux/iphone/bin/ldid: ELF 64-bit LSB shared object, x86-64, version 1 (GNU/Linux), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=2971186340cb12b1a451da2301b69bc85a7b8d2a, for GNU/Linux 3.2.0, not stripped
> ```
> 
> ```
> uname -p -o 
> x86_64 GNU/Linux
> ```
> 
> Not sure what's up. Had a friend test it and it works fine for them (w a 12 sdk as well as 14 sdk). We're both on Ubuntu 20.04 (WSL2).

Same error here on WSL 1.

**Comment by UsrLightmann** [comment]
> > 
> 
> > 
> 
> > > https://github.com/sbingner/llvm-project/releases/tag/v10.0.0-2 should be good for linux with iOS14 SDKs - also updated cctools and ldid
> 
> > 
> 
> > Everything seems good except ldid, which is acting up for me.
> 
> > 
> 
> > ```
> 
> > bash: /home/lightmann/theos/toolchain/linux/iphone/bin/ldid: cannot execute binary file: Exec format error
> 
> > make[2]: *** [/home/lightmann/theos/makefiles/instance/library.mk:51: /home/lightmann/Documents/Tweaks/DEVELOPED/quorra/.theos/obj/Quorra.dylib] Error 126
> 
> > ```
> 
> > 
> 
> > ```
> 
> > file $THEOS/toolchain/linux/iphone/bin/ldid
> 
> > /home/lightmann/theos/toolchain/linux/iphone/bin/ldid: ELF 64-bit LSB shared object, x86-64, version 1 (GNU/Linux), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=2971186340cb12b1a451da2301b69bc85a7b8d2a, for GNU/Linux 3.2.0, not stripped
> 
> > ```
> 
> > 
> 
> > ```
> 
> > uname -p -o 
> 
> > x86_64 GNU/Linux
> 
> > ```
> 
> > 
> 
> > Not sure what's up. Had a friend test it and it works fine for them (w a 12 sdk as well as 14 sdk). We're both on Ubuntu 20.04 (WSL2).
> 
> 
> 
> Same error here on WSL 1. I used the version from 10.0.0-1 to workaround this

After seeing your comment I decided to check my version again...idiot me forgot I went back to WSL1 a little while back. Seems it doesn't play nice with the new ldid. See my edit above.

**Comment by UsrLightmann** [comment]
@Thog After seeing your comment I decided to check my version again...idiot me forgot I went back to WSL1 a little while back. Seems it doesn't play nice with the new ldid. See my edit above.

**Comment by kirb** [comment]
@UsrLightmann Unfortunately I haven’t had time to integrate your work into the wiki, but I’ve just set up a new repo so we can accept pull requests to the wiki. If you have a moment, it’d be really appreciated if you could make a PR on [theos/theos-wiki](https://github.com/theos/theos-wiki). Thanks!

**Comment by UsrLightmann** [comment]
@kirb No worries. Just made a PR for the main Linux/WSL stuff listed above. I don't develop on-device, so I don't know what needs updating in the iOS install section (if anything).

**Comment by kirb** [comment]
Closing this in favor of theos/theos-wiki#1, but I’ll leave it pinned for the time being till that PR is merged.

**Comment by kirb** [comment]
Thanks @UsrLightmann! Updated instructions are now merged in with theos/theos-wiki@1d23b57.
